doc(node): node conditions for environment checks (#947)
longhorn/website: content/docs/1.7.0/nodes-and-volumes/nodes/node-conditions.md

---
title: Node Conditions
weight: 7
---

> Longhorn 1.7.0 introduces node conditions to ensure the environment settings of worker nodes. These conditions help identify and resolve potential issues before they impact the system. Here are the node conditions:

Reviewer: ```suggestion
Node conditions are introduced from Longhorn version 1.7.0. They are used to check the environment settings of worker nodes to identify potential issues before any system impact.

Node conditions:
```

Author: Modified. Thanks.

- `NoMountPropagationSupport`:  

>   Indicates that the node does not support mount propagation, necessary for sharing of volumes mounted by one container with other containers in the same pod of Longhorn, or even to other pods of Longhorn on the same node.  

Reviewer: ```suggestion
  Indicates that the node does not support mount propagation. This is necessary for sharing of volumes mounted by a container with other containers in the same Longhorn pod, or to other Longhorn pods on the same node.  
```

- `MultipathdIsRunning`:  
  Confirms if the `multipathd` service is running on the node, which may affect the pod with the volume startup. See [Troubleshooting: `MountVolume.SetUp failed for volume` due to multipathd on the node](../../../../../kb/troubleshooting-volume-with-multipath).  

- `PackagesNotInstalled`:  
  Identifies if any required packages (`NFS client, iSCSI tool, cryptsetup, device mapper`) for Longhorn are missing.  

- `NFSClientIsNotFound`:  

>   Identifies if all NFS clients (`v4, v4.1, v4.2`) are not enabled, which is required for RWX volume operations.  

Reviewer: ```suggestion
  Identifies if no NFS clients (`v4, v4.1, v4.2`) are enabled. A NFS client is required for RWX volume operations.  
```

Author: Modified. Thanks.

> The conditions will not block the Longhorn deployment but they will result in warnings in the Longhorn `Node` resource.

Reviewer: ```suggestion
Node conditions do not block the Longhorn deployment but they result in warnings in the Longhorn `Node` resource.
```

Author: Modified. Thanks.

---
title: Node Conditions
weight: 7
---

Node conditions are introduced from Longhorn version 1.7.0. They are used to check the environment settings of worker nodes to identify potential issues before any system impact.

Node conditions:

- `NoMountPropagationSupport`:  
  Indicates that the node does not support mount propagation. This is necessary for sharing of volumes mounted by a container with other containers in the same Longhorn pod, or to other Longhorn pods on the same node.  

- `MultipathdIsRunning`:  
  Confirms if the `multipathd` service is running on the node, which may affect the pod with the volume startup. See [Troubleshooting: `MountVolume.SetUp failed for volume` due to multipathd on the node](../../../../../kb/troubleshooting-volume-with-multipath).  

- `PackagesNotInstalled`:  
  Identifies if any required packages (`NFS client, iSCSI tool, cryptsetup, device mapper`) for Longhorn are missing.  

- `NFSClientIsNotFound`:  
  Identifies if no NFS clients (`v4, v4.1, v4.2`) are enabled. A NFS client is required for RWX volume operations. 

Node conditions do not block the Longhorn deployment but they result in warnings in the Longhorn `Node` resource.

> For more information, see [Longhorn Installation Requirements](../../../deploy/install/#installation-requirements).

Reviewer 2: Some conditions are not mentioned
```
const (
	NodeConditionReasonManagerPodDown            = "ManagerPodDown"
	NodeConditionReasonManagerPodMissing         = "ManagerPodMissing"
	NodeConditionReasonKubernetesNodeGone        = "KubernetesNodeGone"
	NodeConditionReasonKubernetesNodeNotReady    = "KubernetesNodeNotReady"
	NodeConditionReasonKubernetesNodePressure    = "KubernetesNodePressure"
	NodeConditionReasonUnknownNodeConditionTrue  = "UnknownNodeConditionTrue"
	NodeConditionReasonNoMountPropagationSupport = "NoMountPropagationSupport"
	NodeConditionReasonMultipathdIsRunning       = "MultipathdIsRunning"
	NodeConditionReasonUnknownOS                 = "UnknownOS"
	NodeConditionReasonNamespaceExecutorErr      = "NamespaceExecutorErr"
	NodeConditionReasonPackagesNotInstalled      = "PackagesNotInstalled"
	NodeConditionReasonKernelConfigIsNotFound    = "KernelConfigIsNotFound"
	NodeConditionReasonNFSClientIsNotFound       = "NFSClientIsNotFound"
	NodeConditionReasonKubernetesNodeCordoned    = "KubernetesNodeCordoned"
)

```

---
title: Node Conditions
weight: 7
---

> Node conditions are introduced from Longhorn version 1.7.0. They are used to check the environment settings of worker nodes to identify potential issues before any system impact.

Reviewer 2: Some conditions are not introduced since v1.7.0.

> Node conditions are introduced from Longhorn version 1.7.0 (`Multipathd`, `RequiredPackages` and `NFSClientInstalled`). They are used to check the environment settings of worker nodes to identify potential issues before any system impact.

Reviewer 2: No need to mention any version and specific conditions in the paragraph. It is simply describing the purpose of node conditions. For example: https://kubernetes.io/docs/reference/node/node-status/#condition

Node conditions:

- `Ready`:  
  Indicates that the node is ready for Longhorn operations, including that a `longhorn-manager` pod is running on this node, the Kubernetes node is ready, and there is no physical resources pressure.  

- `Schedulable`:  
  Indicated that the node is not cordoned and workload can be scheduled to this node.

- `MountPropagation`:  
  Indicates that the node supports mount propagation. This is necessary for sharing of volumes mounted by a container with other containers in the same Longhorn pod, or to other Longhorn pods on the same node.  

- `Multipathd`:  
  Confirms if the `multipathd` service is not running on the node, which may affect the pod with the volume startup. See [Troubleshooting: `MountVolume.SetUp failed for volume` due to multipathd on the node](../../../../../kb/troubleshooting-volume-with-multipath).  

> - `RequiredPackages`:  

Reviewer 3: Will this condition be updated when the data engine changes? for example, v1/v2 exclusive or both enabled? Or right now, we just design this for v1 only, and will be improved in the future?  

cc @derekbit

Reviewer 2: Yes, we have a [ticket](https://github.com/longhorn/longhorn/issues/7874) for v2 data engine as well. You're right. We need to check the packages and dependencies based on the enabled data engines.

Reviewer 3: Yeah. Just make sure the check respecting the data engine when v2 is ready.

>   Identifies if any required packages (`NFS client, iSCSI tool, cryptsetup, device mapper`) for Longhorn are not missing.  

Reviewer 2: Suggest adding link to the installation page for each package.
```suggestion
  Checks if all required packages (NFS client, iSCSI tool, cryptsetup, dmsetup) exist for Longhorn
```

- `NFSClientInstalled`:  

>   Identifies if an NFS client (`v4.2, v4.1 or v4`) is enabled. An NFS client is required for RWX volume and backup operations.  

Reviewer 2: ```suggestion
  Identifies if any of the following NFS clients are supported: `v4.2`, `v4.1`, or `v4.0`. NFS client is required for RWX volume and backup.
```

>   Checks if all required packages ([NFS client](../../../deploy/install/#installing-nfsv4-client), [iSCSI tool](../../../deploy/install/#installing-open-iscsi), [cryptsetup](../../../deploy/install/#root-and-privileged-permission), [dmsetup](../../../deploy/install/#using-the-environment-check-script)) exist for Longhorn  

Reviewer 2: I just realized we don't have docs for the installations of dmsetup and cryptsetup. Can you add them as `open-iscsi`?
